r1cs/gadgets/std/src/bits/macros.rs

```diff
@@ -0,0 +1,2532 @@
+                        if cs.get(&bit_gadget_path).is_zero() {
+                            cs.set(&bit_gadget_path, Fr::one());
+                        } else {
+                            cs.set(&bit_gadget_path, Fr::zero());
+                        }
+                        assert!(!cs.is_satisfied());
+
+
+                        // set bit value back
+                        if cs.get(&bit_gadget_path).is_zero() {
+                            cs.set(&bit_gadget_path, Fr::one());
+                        } else {
+                            cs.set(&bit_gadget_path, Fr::zero());
+                        }
+                        assert!(cs.is_satisfied());
+                    }
+
+                    // test with all constants
+                    let num_constraints = cs.num_constraints();
+                    let operands = operand_values.iter().enumerate().map(|(i, val)| {
+                        alloc_fn(&mut cs, format!("alloc constant operand {}", i).as_str(), &VARIABLE_TYPES[0], *val)
+                    }).collect::<Vec<_>>();
+
+                    let result_var = $type_name::mulmany_nocarry(cs.ns(|| "mul constant operands"), &operands).unwrap();
+
+                    test_uint_gadget_value(result_value, &result_var, "sum of constants result correctness");
+                    assert!(cs.is_satisfied());
+                    assert_eq!(num_constraints, cs.num_constraints()); // check that no constraints are added when operands are all constants
+
+                    // check that constraints are not satisfied in case of overflow
+                    let mut cs = ConstraintSystem::<Fr>::new(SynthesisMode::Debug);
+                    let operand_values = (0..NUM_OPERANDS).map(|_| rng.gen_range(max_value..=$native_type::MAX)).collect::<Vec<$native_type>>();
+
+                    let operands = operand_values.iter().enumerate().map(|(i, val)| {
+                        alloc_fn(&mut cs, format!("alloc operand {}", i).as_str(), &VARIABLE_TYPES[i % 3], *val)
+                    }).collect::<Vec<_>>();
+
+
+                    let mut is_overflowing = false;
+                    let result_value: $native_type = operand_values.iter().map(|el| *el).reduce(|a,b| {
+                        let (updated_sum, overflow) = a.overflowing_mul(b);
+                        is_overflowing |= overflow;
+                        updated_sum
+                    }).unwrap();
+                    //check that the multiplication actually overflows, which should always happen given how we generate operand values
+                    assert!(is_overflowing);
+
+                    let result_var = $type_name::mulmany_nocarry(cs.ns(|| "mul overflowing operands"), &operands).unwrap();
+
+                    test_uint_gadget_value(result_value, &result_var, "result of overflowing mul correctness");
+                    assert!(!cs.is_satisfied());
+                }
+
+                #[test]
```

Comment:
Is this test really needed considering that we have separate tests both for mul and add ?

Reply:
Yes, it allows to test conditional arithmetic operations, while previous tests just focus on addmany and mulmany variants. Maybe I can rename it to `test_conditional_nocarry_arithmetic_operations` to make its purpose clearer.